core/src/main/scala/tapir/Validator.scala

```diff
@@ -0,0 +1,56 @@
+package tapir
+
+import tapir.generic.ValidateMagnoliaDerivation
+
+trait Validator[T] { outer =>
+  def validate(t: T): Boolean
+  def contramap[TT](g: TT => T): Validator[TT] = (t: TT) => {
+    outer.validate(g(t))
+  }
+  def forOption: Validator[Option[T]] = (t: Option[T]) => {
+    t.forall(outer.validate)
+  }
+  def forSeq: Validator[Seq[T]] = (t: Seq[T]) => {
+    t.forall(outer.validate)
+  }
+  def forVector: Validator[Vector[T]] = (t: Vector[T]) => {
+    t.forall(outer.validate)
+  }
+  def forList: Validator[List[T]] = (t: List[T]) => {
+    t.forall(outer.validate)
+  }
+  def forArray: Validator[Array[T]] = (t: Array[T]) => {
+    t.forall(outer.validate)
+  }
+  def forSet: Validator[Set[T]] = (t: Set[T]) => {
```

Comment:
maybe we can have a single `ElementValidator` which takes an `Iterable`? The logic is all the same here.

Reply:
I've just added support for collections. There were two problems:
1. I had to create two implementations - one for array and the other one for iterable
2. There is an ugly cast inside each one, types should be ok, but I could omit it.